Use i18next as i18n framework

McGiverGim opened the pull request:
The i18n framework of Google Chrome is limited to some languages:

https://developer.chrome.com/webstore/i18n?csw=1#localeTable

A lot of languages are missing, some are mainstream. To support more locales, this PR changes the i18n by other, a lot more powerful: `i18next`.

This PR gives us three bigs advantages:
- The first is not limit us to a certain number of locales.
- The second is that now, the i18n functions are all wrapped in the `localization.js` file. If in a future we decide to change to another i18n framework (or return to the chrome one) we only will need to modify this file.
- Finally, an external i18n framework, will let us add to the options a combo with the selection of the language to the user, a very demanded feature request. This part has not been done, maybe in a future...

The "disadvantage" since now are:
- We maintain the `messages.json` format of chrome, not using the advanced formats of `i18next`, but for now I don't think they are necessary...
- The `eventPage.js` is loaded before the application, so the `i18next` is not available here, so if some phrase must be shown to the user the chrome i18n will do it, and will use only the languages availables for chrome (fallback to english). By now only the message of new version available to chrome users.

I've made several tests, and all seems to be working, but other tests are welcome.

McGiverGim commented:
Please wait to merge this PR, I've found two strings that are being not translated.

McGiverGim commented:
Solved!!

mikeller commented on gulpfile.js:
```diff
@@ -244,6 +244,8 @@ function dist() {
         './js/libraries/jquery-ui-1.11.4.min.js',
         './js/libraries/d3.min.js',
         './js/libraries/jquery.nouislider.all.min.js',
+        './js/libraries/i18n/i18next.js',
```
Do we have to add copies of these libraries to our repository? I think it would be much preferrable if we leveraged the fact that we are already using `npm` and added them from https://www.npmjs.com/package/i18next .

McGiverGim replied:
I think that the problem is that we are not packaging the node_modules, it is used only to generate the installer, not to execute, and if we change it chrome app does not support the majority of libraries.

mikeller replied:
Yes we are packaging all node modules that are installed as `dependencies`. ;-) There just aren't any configured yet. What is the difference between the i18next library provided by npm and the static file that is contained in this pull request?

McGiverGim commented:
Yes, is better to end moving the files, I will rebase my code after that.

McGiverGim commented:
The one included is part of the module of npm. The modules are not distributed at least in the chrome app. Maybe there is a workaround?

mikeller commented:
If it is part of the `npm` module, we should not just add a static copy of it to the repository.

Why are they not distributed? They are installed into the `dist/` folder, so if they are not distributed something is badly broken and needs to be fixed.

McGiverGim commented:
Maybe my bad? Tomorrow I will recheck it.

mikeller commented:
Seems to be working fine for me:

```
mikeller@pc:~/git/betaflight-configurator$ npm install i18next --save
betaflight-configurator@10.1.0 /home/mikeller/git/betaflight-configurator
└── i18next@10.3.0 

npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.1.3
mikeller@mikeller-pc:~/git/betaflight-configurator$ gulp dist
Building for platform(s): linux64.
[10:02:39] Using gulpfile ~/git/betaflight-configurator/gulpfile.js
[10:02:39] Starting 'dist'...
[10:02:39] Starting 'clean_dist'...
[10:02:39] Finished 'clean_dist' after 84 ms
[10:02:39] Starting 'dist'...
betaflight-configurator@10.1.0 /home/mikeller/git/betaflight-configurator/dist
└── i18next@10.3.0 

[10:02:43] Finished 'dist' after 3.95 s
[10:02:43] Finished 'dist' after 4.03 s
mikeller@pc:~/git/betaflight-configurator$ ls dist/node_modules/
i18next
```

McGiverGim commented:
This is the first part. How to use a module in the chrome app? Using import it something similar?

mikeller commented:
This can take care of it in a relatively seamless fashion:

https://www.npmjs.com/package/chrome-app-module-loader

McGiverGim commented:
Yeah, it seems a good idea. I will test it and post results.

McGiverGim commented:
Here are the result of my tests:
- The `dependencies` work and are packed in the program. 
- The `dependencies` are downloaded each time we make a `gulp dist`, so it is slow for working with it. I think it is because we remove the contents of `dist` folder each time a `gulp` command is released, maybe we must not remove the `dist/node_modules` folder.
- If I reference the `.js` from the `node_modules` folder it works in chrome, like this:
><script type="text/javascript" src="./node_modules/i18next/i18next.js"></script>
- The standalone app does not work if `dependencies` exist. For some reason, the standalone app starts with the default empty nw window, so it NOT works. Maybe some bad configuration that we need to add when packing the app? 
- I tried to add the `chrome-app-module-loader` library, but it didn't work in chrome for me. It throws an error when I made a `require('i18next')`. Maybe I've loaded it wrong, but I've not investigated too much because the standalone app does not work and it seemed more important for me.

So by the moment, something is wrong with dependencies, and I don't know what it can be. @mikeller , @basdelfos some idea?

mikeller commented:
> The dependencies are downloaded each time we make a gulp dist, so it is slow for working with it. I think it is because we remove the contents of dist folder each time a gulp command is released, maybe we must not remove the dist/node_modules folder.

I think we should in general stop to clean all directories before builds - cleaning them only with `make clean` should be sufficient, and it can be annoying when doing things like building multiple different releases in a row, and then finding only the output of the last build was not deleted...

>If I reference the .js from the node_modules folder it works in chrome, like this:
  <script type="text/javascript" src="./node_modules/i18next/i18next.js"></script>
> The standalone app does not work if dependencies exist. For some reason, the standalone app starts with the default empty nw window, so it NOT works. Maybe some bad configuration that we need to add when packing the app?

That's odd. Testing here (linux) with `npm install i18next --save` and adding `<script type="text/javascript" src="./node_modules/i18next/i18next.js"></script>` to `main.js` works fine. Could it be that there is an exception thrown in the initialisation for the translations?

McGiverGim commented:
No, the error happens to me before initialize the translations, before all... When the translations fail, the page is loaded but without tests. 
I will start again, download master, test, add dependency, and test... Maybe I've some rests of an earlier test...

McGiverGim commented:
Strange, I started from scratch and now it works.  The `i18next` and the `backend` are being readed from the `node_modules` directory. I've the other branch, that fails, if I have time I will search what was the problem...

I've updated the PR with the new code.

I will give a try to the `chrome-app-module-loader` but I'm not too confident with it ;)

mikeller commented:
Good work, @McGiverGim!

> I will give a try to the chrome-app-module-loader but I'm not too confident with it ;)

I wasn't sure if it was possible at all to use node modules from within a chrome app without this when I made my recommendation. But as it turns out to be, it works fine without it, so using it will just be 'syntactic sugar'. ;-)

McGiverGim commented:
Yes, my first tests haven't work. I have opened and issue with the author to see if he can helps.
And I have ready the menu to let the user select the language in app 🤗

McGiverGim commented:
Rebased, squashed and solved the problem with the `node_modules` folder introduced here: https://github.com/betaflight/betaflight-configurator/pull/878

I think it is ready to be merged, the modules library for chrome does not work and the user does not help. If we can get it work maybe we can add it in a later stage.

McGiverGim commented:
The same fix has been introduced here: https://github.com/betaflight/betaflight-configurator/pull/888

So maybe is better to merge the fix, then I will rebase this again.

McGiverGim commented:
Rebased against master.
